CoreNetKit/Kit/Safe/AccessSafeContextManager.swift

```diff
@@ -0,0 +1,180 @@
+//
+//  AccessSafeContextManager.swift
+//  CoreNetKit
+//
+//  Created by Александр Кравченков on 09.05.2018.
+//  Copyright © 2018 Кравченков Александр. All rights reserved.
+//
+
+import CoreEvents
+
+// MARK: - Support types
+
+public protocol PerformableContext: class {
+
+    @discardableResult
+    func perform() -> Self
+}
+
+public protocol SafeErrorContext: class {
+
+    @discardableResult
+    func onAccessError(_ completion: @escaping () -> Void) -> Self
+    func performAccessError()
+}
+
+public typealias AccessSafeContext = PerformableContext & SafeErrorContext
+
+/// This object should provide contexts that can
+public protocol RefreshAccessContextProvider {
+    func getContext() -> ActionableContext<Void>
+}
+
+// MARK: - Manager
+
+open class AccessSafeContextManager {
+
+    // MARK: - fileds
+
+    fileprivate var activeContext: [AccessSafeContext]
+
+    fileprivate let semaphore: DispatchSemaphore
+
+    fileprivate var isRefreshTokenRequestWasSended: Bool
+
+    // MARK: - properties
+
+    open var refreshAccessContextProvider: RefreshAccessContextProvider
+
+    public init(refreshAccessContextProvider: RefreshAccessContextProvider) {
+        self.refreshAccessContextProvider = refreshAccessContextProvider
+        self.activeContext = [AccessSafeContext]()
+        self.semaphore = DispatchSemaphore(value: 1)
+        self.isRefreshTokenRequestWasSended = false
+    }
+
+    public func add(context: AccessSafeContext) {
+        DispatchQueue.global(qos: .userInitiated).async {
+            self.backgoundAdd(context: context)
+        }
+    }
+
+    /// Removed all sleeped requests and allows perform new requests
+    public func clear() {
+
+        self.activeContext.removeAll()
+
+        self.semaphore.wait()
+
+        self.isRefreshTokenRequestWasSended = false
+
+        self.semaphore.signal()
+    }
+
+    /// Tried reperform all waiting requests
+    public func update() {
+        self.successSafePerformation()
+    }
+}
+
+// MARK: - Backround methods
+
+private extension AccessSafeContextManager {
+
+    func backgoundAdd(context: AccessSafeContext) {
+
+        self.semaphore.wait()
+
+        // true if no need perform
+        // false if perform needed
+        guard !isRefreshTokenRequestWasSended else {
+            self.semaphore.signal()
+            return
+        }
+
+        self.semaphore.signal()
```

Comment:
You're calling `self.semaphore.signal()` twice in code, but you could call once before guard body

Reply:
And maybe there is no need to use the `semaphore`?

Reply:
No. It in place.